K2Field.K2NE.ServiceBroker/ServiceObjects/ExcelServices/ExcelImportServiceSO.cs

```diff
@@ -0,0 +1,528 @@
+
+            DataRow dr = results.NewRow();
+            dr[Constants.SOProperties.ExcelImportServices.Results] = result;
+
+            results.Rows.Add(dr);
+        }
+
+        /// <summary>
+        /// Import data from Excel file into SmartObject
+        /// </summary>
+        private string Import(FileProperty excelFile, string sheetName, string smartObject, string createMethodName, string headerRowSpaces, string transactionIDName,
+            string transactionIDValue, string FQN)
+        {
+            string result = string.Empty;
+
+            byte[] bFile = Convert.FromBase64String(excelFile.Content);
+            MemoryStream stream = new MemoryStream(bFile);
+
+            // arrays to store the imported column names and also the SmartObject column names
+            // this will be used later to get a list of matching columns.
+            string[] dsColumnNames = null, soColumnNames = null;
+
+            // To store returned values from Excel file
+            DataTable dt;
+
+            // Read Data in excel file
+            try
+            {
+                dt = ReadExcelFile(stream, sheetName);
+
+                if (dt.Rows.Count == 0)
+                {
+                    result = "0 rows imported";
+                    return result;
+                }
+                else
+                {
+                    result = dt.Rows.Count.ToString() + " rows found. ";
+                    if (dt.Columns.Count == 0)
+                    {
+                        result = "0 columns found.";
+                        return result;
+                    }
+                    else
+                    {
+                        // populate an array of column names
+                        dsColumnNames = new string[dt.Columns.Count];
+                        foreach (DataColumn col in dt.Columns)
+                        {
+                            if (headerRowSpaces != null && headerRowSpaces.Trim().ToLower() == "remove")
+                            {
+                                col.ColumnName = col.ColumnName.Replace(" ", "");
+                            }
+                            else  // by default replace spaces with underscores as SmartObject system names do that
+                            {
+                                col.ColumnName = col.ColumnName.Replace(" ", "_");
+                            }
+                            // just get rid of other (non-underscore or hyphen) punctuation which is also invalid
+                            col.ColumnName = Regex.Replace(col.ColumnName, @"[\p{P}\p{S}-[-_]]", "");
+                            dsColumnNames[col.Ordinal] = col.ColumnName.ToLower();
+                        }
+                        result += "Columns found: " + string.Join(",", dsColumnNames);
+                    }
+                }
+            }
+            catch (IOException ex)
```

Comment:
See other note on exceptin handling. I'd suggest to not handle this and bubble it up to the top or you provide an error message and full details of the actual exception.

Reply:
Done.